---
title: "Further Statistical Analysis using R"
author: "Mark Dunning, Matthew Eldridge and Sarah Vowler"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
output:  BiocStyle::pdf_document
toc: TRUE
---

# Introduction

```{r eval=TRUE, echo=F, results="asis"}
BiocStyle::markdown()

```



In this practical, we will use several ’read-life’ datasets to demonstrate some of the concepts you have
seen in the lectures. We will guide you through how to analyse these datasets in R and the kinds
of questions you should be asking yourself when faced with similar data. As this is not a course in using R, we will provide the R code that you will need. However, it is up to you to think carefully about the assumptions of a statistics tests and interpret the results that R gives

```{r knitr, echo=FALSE, results="hide"}
library("knitr")
opts_chunk$set(tidy=FALSE,dev="png",fig.show="as.is",
               fig.width=10,fig.height=4,
               message=FALSE,eval=TRUE,warning=FALSE,echo=TRUE)
```

The data you need for this practical are available as a [https://rawgit.com/bioinformatics-core-shared-training/intermediate-stats/master/CourseData.zip](zip file). Please download this zip file and extract to a directory on your laptop. You will then need to tell RStudio where to find these files by setting your *working directory*

Session -> Set Working Directory -> Choose Directory

# One-Way ANOVA

The data for this exercise are to be found in `amess.csv`. The data are the red cell folate levels in three groups of cardiac bypass patients given different levels of nitrous oxide (N$_2$O) and oxygen (O$_2$) ventilation. [There is a reference to the source of this data in Altman, Practical Statistics for Medical Research, p. 208.]

The treatments are

i) 50% N2O and 50% O2 continuously for 24 hours
ii) 50% N2O and 50% O2 during the operation
iii) No N2O but 35-50% O2 continuously for 24 hours

1. Import the file `amess.csv` into R.  Verify that the dimensions of the object that R creates are correct.

```{r}
amess <- read.csv("amess.csv")
dim(amess)
```

***NOTE: You can use the `View` function in RStudio to bring-up a display of the object you have created***

2. Draw boxplots of the data. Does it look as though the assumptions for ANOVA are satisfied?

***NOTE: It is often useful to overlay individual points on the boxplot***

```{r}
library(beeswarm)
boxplot(folate~treatmnt,data=amess)
beeswarm(folate~treatmnt,data=amess,add=TRUE,pch=16)
```

3. Perform t-tests for each of the three pair-wise comparisons. Make note of the t-test statistics obtained. Which groups, if any, differ from one another?

***NOTE:If you have not performed a t test in R before, you may wish to consult the help page for `t.test`; `?t.test`. ***


```{r}
t.test(folate~ treatmnt,data=amess[amess$treatmnt !=3,])
t.test(folate~treatmnt,data=amess[amess$treatmnt !=2,])
t.test(folate~treatmnt,data=amess[amess$treatmnt !=3,])
```


4. Perform an analysis of variance on the data. Why is this a more valid analysis? Interpret the output.


```{r}
mod <- aov(folate~factor(treatmnt),data=amess)
mod
summary(aov(mod))
```


5. Choose any pair of treatments and repeat the ANOVA for just these two treatments. Compare the test statistic value to the results of the t-test from question 3. What do you notice?


```{r}
subset <- amess[amess$treatmnt != 3,]
t.test(folate~treatmnt,data=subset)
```


6. If the assumptions for ANOVA don’t hold, then there are non-parametric alternatives available. 

    + If the assumption of normality doesn’t hold then we might look at the Kruskal-Wallis test. 
    + If the assumption of equal variances doesn’t hold then we might use the median test.

Perform these tests. Do they give the same answer as ANOVA? If not, why not? How important were the assumptions?


```{r}
kruskal.test(folate~factor(treatmnt),data=amess)

library(RVAideMemoire)
mood.medtest(folate~factor(treatmnt),data=amess)
```

7. Use a *post-hoc* test to compute p-value for all pairwise contrasts. Unlike the t-tests performed in Question 3, the p-values returned are adjusted for multiple testing. Two functions for doing such tests in R are `TukeyHSD` and `pairwise.t.test`. Help for both these functions is available through RStudio; `?TukeyHSD`, `?pairwise.t.test`

```{r}
TukeyHSD(mod)
```


```{r}
pairwise.t.test(amess$folate,amess$treatmnt)
```


# Other Non-Parametric tests

The file `treatments.txt` records measurements from five subjects undergoing a treatment course. The data were taken from an example in *GraphPad Prism*.

1. Import these data into R

```{r}
treatments <- read.delim("treatments.txt")
head(treatments)
```

2. Choose an appropriate method to test the effect of treatment. Interpret the result.

```{r}
boxplot(treatments[,-1])
friedman.test(as.matrix(treatments[,-1]))
```


```{r echo=FALSE}
##make up some genotype-expression association
if(!file.exists("genotypes.txt")){
  set.seed("06102015")
  AA <- c(rnorm(5,mean = 6,sd = 2),NA,NA)
  AB <- rnorm(7,mean=7,sd=2)
  BB <- c(NA,rnorm(6, mean=8,sd=1))
  
  gtypes <- data.frame(AA,AB,BB)
  write.table(gtypes, "genotypes.txt",sep="\t")
}

```

The file `genotypes.txt` contains measurements from a gene expression study. For each patient in the study, their genotype for a particular gene was recorded. 

3. Import the file `genotypes.txt` into R

```{r}
gtypes <- read.delim("genotypes.txt")
gtypes
```

4. Transform the data into two columns; one to indicate the genotype of the individual and the second for the gene expression level. 


```{r}
library(tidyr)
gtypes <- gather(gtypes)
gtypes$key <- rep(c(1,2,3),each=7)
```

5. Use an appropriate method to test the association between gene expression and genotype

```{r}
library(clinfun)
jonckheere.test(gtypes$value,gtypes$key)
```



# Breast Cancer Incidence

The file `globalBreastCancerRisk.csv` gives the number of new cases of Breast Cancer (per population of 10,000) in various countries around the world, along with various health and lifestyle risk factors. These data were collected from the [gapminder](www.gapminder.org) resource \footnote{Similar data are also detailed in a blog post at: http://www.analyticsforfun.com/2014/06/performing-anova-test-in-r-results-and.html}. Let's suppose we are initially interested in whether the number of breast cancer cases is significantly different in different regions of the world.

1. Read these data into R
    
```{r}
bcStats <- read.csv("globalBreastCancerRisk.csv")
head(bcStats)
```

    
2.  Visualise the distribution of breast cancer incidence in each continent. Include a way of identifying how many observations belong to each group (continent)

```{r}
library(RColorBrewer)
library(beeswarm)
boxplot(NewCasesOfBreastCancerIn2002~continent,data=bcStats,col=brewer.pal(5,"Set1"),notch=TRUE)
beeswarm(NewCasesOfBreastCancerIn2002~continent,data=bcStats,add=TRUE,pch=16)
bcStats.clean <- bcStats[bcStats$continent != "Oceania",]
```

3. Would a parameteric, or non-parametric, approach be suitable for this analysis? Use the result from a `bartlett.test` to support you answer
Proceed with your chosen approach to test the hypothesis that breast cancer incidence is different across the globe

```{r}
bartlett.test(NewCasesOfBreastCancerIn2002~continent,data=bcStats.clean)
kruskal.test(NewCasesOfBreastCancerIn2002~continent,data=bcStats.clean)
```

4. In a manner similar to the *One-Way ANOVA* example, we can apply multiple-testing correction to all pairwise contrasts and assess which individual contrasts are significant. Which pairs of continents have significantly different breast cancer incidence rates to each other?
    
```{r}
pairwise.wilcox.test(bcStats.clean$NewCasesOfBreastCancerIn2002,bcStats.clean$continent)
```
    
# Linear Regression

1. Clinical Trial Data: A researcher wished to study the effect of a particular drug X on the immune system. Thus, the concentration of the drug was measured on various days, along with the white blood cell counts.

    a) Import the clinical data from the file `ClinicalTrials.txt` 
    b) Make a scatter plot
    c) Label each point according to the date that the measurement was made
    d) Fit a linear model to the data and produce the diagnostic plots. Which observations fit the model least-well? Compare your answers to the scatter plot
    e) Overlay the line-of-best fit on the scatter plot. Indicate the differences between each fitted and observed value. 
    f) Extract the $R^2$ value from the model summary, and print this value on the plot
    
```{r fig.height=5,fig.width=5}
trials <- read.delim("clinicalTrials.txt")

x <- trials$Drug.concentration
y <- trials$Cell.Count

plot(x,y,cex=0.8)
text(x,y+0.2,labels=1:length(x))
```

***NOTE: We use the `text` function to overlay the number of each observation on-top of the scatter plot that we create initially. `text` requires a set of x and y coordinates. Here we add an offset of 0.2 to make sure the labels are not obscured by the points***

```{r fig.height=6,fig.width=6}
model <- lm(y ~x )
summary(model)
par(mfrow=c(2,2))
plot(model)
```

***NOTE: We modify the plot layout by changing the mfrow parameter to allow for all the diagnostic plots to be shown in the same plotting device***

```{r}
par(mfrow=c(1,1))
barplot(model$residuals,names=1:length(x),las=2)
```

***NOTE: The setting of the `las` parameter to 2 arranges the labels so they are perpendicular to the x-axis***

```{r,fig.height=5,fig.width=7}
par(mfrow=c(1,1))
plot(x,y,pch=16)
points(x,model$fitted.values,pch=16,col="grey")
arrows(x,model$fitted.values,x,y,code=0,col="grey")
abline(model,col="red",lty=2)
text(x,y+0.2,labels=1:length(x),cex=0.8,col="steelblue")
```


***NOTE: The `arrows` function was used to draw vertical lines on the plot. Each line is defined by a set of starting and ending x and y positions. In this case, we want the start and end positions to be the same and to vary the y coordinates***

```{r}
plot(x,y,pch=16)
abline(model,col="red",lty=2)
result <- summary(model)
r2 <- round(result$adj.r.squared,3)
my.p = result$coefficients[2,4]

mylabel = bquote(italic(R)^2 == .(format(r2, digits = 3)))
text(x = 5000, y = 8, labels = mylabel)

```

2. Interpolating Unknown values

```{r}
assay <- read.delim("Assay.txt")
model <- lm(micrograms~Optical.Density,data=assay)
plot(assay$Optical.Density, assay$micrograms)
abline(model)
```


```{r}
y <- assay$micrograms
x <- assay$Optical.Density

topredict <- which(is.na(y))
new <- data.frame(x = x[which(is.na(y))])
y[which(is.na(y))] <- predict(lm(y~x), new)

```

```{r}
cols <- rep("black",length(x))
cols[topredict] <- "red"
plot(x,y,pch=16,col=cols)
```

